### ISSUE TITLE
task reset from succeeded to running

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-flow#6315]
There seems to be some circumstance where succeeded tasks can go back to running (apparently in the same main-loop iteration).

This log is typical of the issue:

```console
$ grep 20190401T0000Z/archive_logs log/scheduler/log
2024-07-29T22:59:59Z INFO - [20190401T0000Z/archive_logs:waiting(runahead)] => waiting
2024-07-30T13:17:30Z INFO - [20190401T0000Z/archive_logs:waiting] => waiting(queued)
2024-07-30T13:17:30Z INFO - [20190401T0000Z/archive_logs:waiting(queued)] => waiting
2024-07-30T13:17:31Z INFO - [20190401T0000Z/archive_logs:waiting] => preparing
2024-07-30T13:17:43Z INFO - [20190401T0000Z/archive_logs/01:preparing] submitted to xce:pbs[2492608]
2024-07-30T13:17:44Z INFO - [20190401T0000Z/archive_logs/01:preparing] => submitted
2024-07-30T13:18:05Z INFO - [20190401T0000Z/archive_logs/01:submitted] => running
2024-07-30T13:18:39Z INFO - [20190401T0000Z/archive_logs/01:running] => succeeded
2024-07-30T13:18:39Z INFO - [20190401T0000Z/archive_logs/01:running(runahead)] => running
```

Every time we see the confusing `running(runahead) => running` transition. There are no obvious exacerbating circumstances.

This seems to be intermittent, but we appear to have an example which yields this bug relatively regularly, though I've not been able to reproduce it yet.

**Comment by hjoliver** [comment]
It might be useful to see the whole log, not just lines associated with the affected task.  Maybe there was some kind of manual intervention that affected runahead tasks?  e.g. triggering a runahead task. I'm not sure that we're properly removing task attribute labels like "queued" and "runahead" every time when we (e.g.) force a runahead-limited task task to run.

**Comment by oliver-sanders** [comment]
> It might be useful to see the whole log, not just lines associated with the affected task

The whole log is very long. I went through carefully and could not find any other relevant log lines.

* There were no client commands at all.
* There were no restarts/reloads.
* There were no unexpected external factors or failures.

**Comment by oliver-sanders** [comment]
I haven't managed to make head or tail of this one yet. One user's workflows have encountered this a few times, they are on leave at the moment, hopefully when they return we can get them to run these workflows in debug mode which might give us a better chance of debugging.

**Comment by oliver-sanders** [comment]
I have observed this with a simple workflow whilst testing something else on 8.4.0.dev (I can't say exactly what branch I had checked out at the time, not that it makes a difference).


<details>
<summary>log/config/01-start-01.cylc</summary>

```cylc
# cylc-version: 8.4.0.dev                                                                                                          
[scheduler]                                                                 
    UTC mode = True
    allow implicit tasks = True                                        
[scheduling]                                           
    initial cycle point = 2019-12-09T09:00Z   
    [[graph]]
        R1 = foo_cold => foo_start               
        R1/T00 = foo_start[^] => FOO
        T00, T12 = """
            cycle_end[-PT12H] => FOO
            FOO:succeed-all => BAR
            BAR:succeed-any => wipe_bar
            BAR:succeed-all & wipe_bar => cycle_end
        """
[runtime]
    [[root]]
        script = sleep 2
    [[FOO]]
        script = sleep 2
    [[BAR]]
        script = sleep 2
    [[foo]]
        script = sleep 2
        inherit = FOO
        completion = succeeded
    [[foot]]
        script = sleep 2
        inherit = FOO
        completion = succeeded
    [[fool]]
        script = sleep 2
        inherit = FOO
        completion = succeeded
    [[bar]]
        script = sleep 2
        inherit = BAR
        completion = succeeded
    [[bool]]
        script = sleep 2
        inherit = BAR
        completion = succeeded
    [[pub]]
        script = sleep 2
        inherit = BAR
        completion = succeeded
    [[foo_cold]]
        completion = succeeded
    [[foo_start]]
        completion = succeeded
    [[wipe_bar]]
        completion = succeeded
    [[cycle_end]]
        completion = succeeded
```

</details>

<details>

<summary>log/scheduler/log</summary>

```log
2024-10-14T09:12:49Z INFO - Workflow: generic/run6    
2024-10-14T09:12:50Z INFO - Scheduler: url=tcp://myhost:43244 pid=116508    
2024-10-14T09:12:50Z INFO - Workflow publisher: url=tcp://myhost:43197                                                             
2024-10-14T09:12:50Z INFO - Run: (re)start number=1, log rollover=1    
2024-10-14T09:12:50Z INFO - Cylc version: 8.4.0.dev    
2024-10-14T09:12:50Z INFO - Run mode: live    
2024-10-14T09:12:50Z INFO - Initial point: 20191209T0900Z    
2024-10-14T09:12:50Z INFO - Final point: None    
2024-10-14T09:12:50Z INFO - Cold start from 20191209T0900Z    
2024-10-14T09:12:50Z INFO - New flow: 1 (original flow from 20191209T0900Z) 2024-10-14T09:12:50+00:00    
2024-10-14T09:12:50Z INFO - [20191209T1200Z/foo:waiting(runahead)] => waiting    
2024-10-14T09:12:50Z INFO - [20191209T1200Z/foo:waiting] => waiting(queued)    
2024-10-14T09:12:50Z INFO - [20191209T0900Z/foo_cold:waiting(runahead)] => waiting    
2024-10-14T09:12:50Z INFO - [20191209T0900Z/foo_cold:waiting] => waiting(queued)    
2024-10-14T09:12:50Z INFO - [20191209T1200Z/fool:waiting(runahead)] => waiting    
2024-10-14T09:12:50Z INFO - [20191209T1200Z/fool:waiting] => waiting(queued)    
2024-10-14T09:12:50Z INFO - [20191209T1200Z/foot:waiting(runahead)] => waiting    
2024-10-14T09:12:50Z INFO - [20191209T1200Z/foot:waiting] => waiting(queued)    
2024-10-14T09:12:50Z INFO - Pausing the workflow: Paused on start up    
2024-10-14T09:12:59Z INFO - Command "hold" received. ID=36df3e8e-2284-4149-bb78-538d5fba4e4a    
    hold(tasks=['*'])    
2024-10-14T09:12:59Z INFO - [20191209T1200Z/foo:waiting(queued)] => waiting(held)(queued)    
2024-10-14T09:12:59Z INFO - [20191209T1200Z/fool:waiting(queued)] => waiting(held)(queued)    
2024-10-14T09:12:59Z INFO - [20191209T1200Z/foot:waiting(queued)] => waiting(held)(queued)    
2024-10-14T09:12:59Z INFO - [20191209T0900Z/foo_cold:waiting(queued)] => waiting(held)(queued)    
2024-10-14T09:12:59Z INFO - Command "hold" actioned. ID=36df3e8e-2284-4149-bb78-538d5fba4e4a    
2024-10-14T09:13:17Z INFO - Command "resume" received. ID=fc295c90-f4d6-4f3e-bf50-b09350fcaebd    
    resume()    
2024-10-14T09:13:17Z INFO - RESUMING the workflow now    
2024-10-14T09:13:17Z INFO - Command "resume" actioned. ID=fc295c90-f4d6-4f3e-bf50-b09350fcaebd    
2024-10-14T09:13:44Z INFO - Command "hold" received. ID=1a2c5c93-9e3c-4710-8e46-a367b86db2a7    
    hold(tasks=['20191209T1200Z/bar'])      
2024-10-14T09:13:45Z INFO - Command "hold" actioned. ID=1a2c5c93-9e3c-4710-8e46-a367b86db2a7    
2024-10-14T09:13:54Z INFO - Command "hold" received. ID=84a9070b-2415-48cc-ab35-f4e378c521d6    
    hold(tasks=['20191209T1200Z/bool'])     
2024-10-14T09:13:54Z INFO - Command "hold" actioned. ID=84a9070b-2415-48cc-ab35-f4e378c521d6    
2024-10-14T09:14:00Z INFO - Command "hold" received. ID=2f0629ea-efc4-4fd8-8518-cc87856c8193    
    hold(tasks=['20191209T1200Z/pub'])      
2024-10-14T09:14:00Z INFO - Command "hold" actioned. ID=2f0629ea-efc4-4fd8-8518-cc87856c8193    
2024-10-14T09:14:11Z INFO - Command "force_trigger_tasks" received. ID=5b3a6d05-a651-45c5-a04e-d8281a9d0e17    
    force_trigger_tasks(flow=['all'], flow_wait=False, tasks=['20191209T1200Z/pub'])    
2024-10-14T09:14:11Z INFO - [20191209T1200Z/pub:waiting(runahead)] => waiting    
2024-10-14T09:14:11Z INFO - [20191209T1200Z/pub:waiting] => waiting(queued)    
2024-10-14T09:14:11Z INFO - Command "force_trigger_tasks" actioned. ID=5b3a6d05-a651-45c5-a04e-d8281a9d0e17    
2024-10-14T09:14:11Z INFO - [20191209T1200Z/pub:waiting(queued)] => waiting    
2024-10-14T09:14:11Z INFO - [20191209T1200Z/pub:waiting] => preparing    
2024-10-14T09:14:13Z INFO - [20191209T1200Z/pub/01:preparing] submitted to localhost:background[117362]    
2024-10-14T09:14:13Z INFO - [20191209T1200Z/pub/01:preparing] => submitted    
2024-10-14T09:14:16Z INFO - [20191209T1200Z/pub/01:submitted] => running    
2024-10-14T09:14:20Z INFO - [20191209T1200Z/pub/01:running] => succeeded    
2024-10-14T09:14:20Z INFO - [20191209T1200Z/wipe_bar:waiting(runahead)] => waiting    
2024-10-14T09:14:20Z INFO - [20191209T1200Z/wipe_bar:waiting] => waiting(queued)    
2024-10-14T09:14:20Z INFO - [20191209T1200Z/cycle_end:waiting(runahead)] => waiting    
2024-10-14T09:14:20Z INFO - Command "force_trigger_tasks" received. ID=c4a0cff6-0a89-487b-b5cb-67d2aaf5f2d0    
    force_trigger_tasks(flow=['all'], flow_wait=False, tasks=['20191209T1200Z/bool'])    
2024-10-14T09:14:21Z INFO - [20191209T1200Z/bool:waiting(runahead)] => waiting    
2024-10-14T09:14:21Z INFO - [20191209T1200Z/bool:waiting] => waiting(queued)    
2024-10-14T09:14:21Z INFO - Command "force_trigger_tasks" actioned. ID=c4a0cff6-0a89-487b-b5cb-67d2aaf5f2d0    
2024-10-14T09:14:21Z INFO - [20191209T1200Z/wipe_bar:waiting(queued)] => waiting    
2024-10-14T09:14:21Z INFO - [20191209T1200Z/bool:waiting(queued)] => waiting    
2024-10-14T09:14:21Z INFO - [20191209T1200Z/bool:waiting] => preparing    
2024-10-14T09:14:21Z INFO - [20191209T1200Z/wipe_bar:waiting] => preparing    
2024-10-14T09:14:22Z INFO - [20191209T1200Z/bool/01:preparing] submitted to localhost:background[117722]    
2024-10-14T09:14:22Z INFO - [20191209T1200Z/bool/01:preparing] => submitted    
2024-10-14T09:14:22Z INFO - [20191209T1200Z/wipe_bar/01:preparing] submitted to localhost:background[117723]    
2024-10-14T09:14:22Z INFO - [20191209T1200Z/wipe_bar/01:preparing] => submitted    
2024-10-14T09:14:26Z INFO - [20191209T1200Z/wipe_bar/01:submitted] => running    
2024-10-14T09:14:26Z INFO - [20191209T1200Z/bool/01:submitted] => running    
2024-10-14T09:14:28Z INFO - [20191209T1200Z/wipe_bar/01:running] => succeeded    
2024-10-14T09:14:28Z INFO - [20191209T1200Z/bool/01:running] => succeeded    
2024-10-14T09:14:28Z INFO - [20191209T1200Z/wipe_bar/01:running(runahead)] => running    
2024-10-14T09:14:34Z INFO - Command "force_trigger_tasks" received. ID=4ee0bfc8-2b14-42e8-81f3-27ba635c5930    
    force_trigger_tasks(flow=['all'], flow_wait=False, tasks=['20191209T1200Z/foo'])    
2024-10-14T09:14:34Z INFO - Command "force_trigger_tasks" actioned. ID=4ee0bfc8-2b14-42e8-81f3-27ba635c5930    
2024-10-14T09:14:34Z INFO - [20191209T1200Z/foo:waiting(held)(queued)] => waiting(held)    
2024-10-14T09:14:34Z INFO - [20191209T1200Z/foo:waiting(held)] => preparing(held)    
2024-10-14T09:14:36Z INFO - [20191209T1200Z/foo/01:preparing(held)] submitted to localhost:background[118435]    
2024-10-14T09:14:36Z INFO - [20191209T1200Z/foo/01:preparing(held)] => submitted(held)    
2024-10-14T09:14:40Z INFO - [20191209T1200Z/foo/01:submitted(held)] => running(held)    
2024-10-14T09:14:42Z INFO - [20191209T1200Z/foo/01:running(held)] => succeeded(held)    
2024-10-14T09:14:42Z INFO - [20191209T1200Z/bar:waiting(runahead)] holding (as requested earlier)    
2024-10-14T09:14:42Z INFO - [20191209T1200Z/bar:waiting(runahead)] => waiting(held)(runahead)    
2024-10-14T09:14:42Z INFO - [20191209T1200Z/bar:waiting(held)(runahead)] => waiting(held)    
2024-10-14T09:14:42Z INFO - [20191209T1200Z/bool:succeeded] already finished and completed (flows=1))    
2024-10-14T09:14:42Z INFO - [20191209T1200Z/pub:succeeded] already finished and completed (flows=1))    
2024-10-14T09:14:49Z INFO - Command "stop" received. ID=d412b0a7-b812-41c2-aad0-6091fd1cd9b7    
    stop(mode=Clean)    
2024-10-14T09:14:50Z INFO - Command "stop" actioned. ID=d412b0a7-b812-41c2-aad0-6091fd1cd9b7
```

</details>

The result is a workflow stuck in the "stopping" state with the status message "stopping: waiting for active jobs to complete".

The affected task has the characteristic `running(runhead) => running` message:

```console
$ cylc cat generic | grep wipe_bar
2024-10-14T09:14:20Z INFO - [20191209T1200Z/wipe_bar:waiting(runahead)] => waiting
2024-10-14T09:14:20Z INFO - [20191209T1200Z/wipe_bar:waiting] => waiting(queued)
2024-10-14T09:14:21Z INFO - [20191209T1200Z/wipe_bar:waiting(queued)] => waiting
2024-10-14T09:14:21Z INFO - [20191209T1200Z/wipe_bar:waiting] => preparing
2024-10-14T09:14:22Z INFO - [20191209T1200Z/wipe_bar/01:preparing] submitted to localhost:background[117723]
2024-10-14T09:14:22Z INFO - [20191209T1200Z/wipe_bar/01:preparing] => submitted
2024-10-14T09:14:26Z INFO - [20191209T1200Z/wipe_bar/01:submitted] => running
2024-10-14T09:14:28Z INFO - [20191209T1200Z/wipe_bar/01:running] => succeeded
2024-10-14T09:14:28Z INFO - [20191209T1200Z/wipe_bar/01:running(runahead)] => running
```

**Comment by oliver-sanders** [comment]
My example reproduced a couple of times, but then stopped reproducing :(

It is heavily timings dependent, I'm having more luck with this example at the moment:

```bash
#!/usr/bin/env bash

poll () {
    FP=$(cylc cat generic -m p)
    for i in $(seq 0 100); do
        if grep "$1" "$FP"; then
            return
        fi
        sleep 0.1
    done
}

cylc vip generic --pause; sleep 2

cylc hold 'generic//*'; sleep 1

cylc play generic; sleep 3

cylc hold generic//20191209T1200Z/bar;
cylc hold generic//20191209T1200Z/bool;
cylc hold generic//20191209T1200Z/pub;

cylc trigger generic//20191209T1200Z/pub
# cylc workflow-state generic//20191209T1200Z/pub:running
poll 'cycle_end:waiting(runahead)\] => waiting'
cylc trigger generic//20191209T1200Z/bool
cylc trigger generic//20191209T1200Z/foo
```

**Comment by hjoliver** [comment]
(Good job tracking that down)

**Comment by oliver-sanders** [comment]
Closed by #6409
